File: customview/src/main/java/com/example/customview/activity/HoverItemActivity2.java
package com.example.customview.activity;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.TextView;

import com.example.customview.R;
import com.example.customview.adapter.HoverItemAdapter2;

import java.util.ArrayList;
import java.util.List;

public class HoverItemActivity2 extends AppCompatActivity {
    private static final String TAG = "MainActivity";
    private HoverItemAdapter2 mAdapter;
    private RecyclerView mRecyclerView;
    private List<String> mDatas;
    private View mIncludeView;
    private String mTypeName;
    //目标项是否在最后一个可见项之后
    private boolean mShouldScroll;
    //记录目标项位置
    private int mToPosition;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_hover_item_2);
        mIncludeView = findViewById(R.id.include);
        mRecyclerView = findViewById(R.id.recyclerView);
        //添加数据
        addData();
        mAdapter = new HoverItemAdapter2(this, mDatas);
        mRecyclerView.setAdapter(mAdapter);
        //线性layout
        mRecyclerView.setLayoutManager(new LinearLayoutManager(this));
        //滑动监听
        mRecyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
                //定位初始点坐标
                View stickyInfoView = recyclerView.findChildViewUnder(recyclerView.getMeasuredWidth() >> 1, 1);
                if (stickyInfoView != null) {
                    TextView typeName = stickyInfoView.findViewById(R.id.typeName);
                    if (typeName != null) {
                        mTypeName = typeName.getText().toString();
                    }
                    Log.i(TAG, "onScrolled: mTypeName = " + mTypeName);
                    if (mTypeName.equals("第一项")) {
                        mIncludeView.setVisibility(View.GONE);
                    } else if (mTypeName.equals("第二项")) {
                        mIncludeView.setVisibility(View.VISIBLE);
                    } else if (mTypeName.equals("第三项")) {
                        mIncludeView.setVisibility(View.VISIBLE);
                    }
                }
            }

            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                if (mShouldScroll && RecyclerView.SCROLL_STATE_IDLE == newState) {
                    mShouldScroll = false;
                    smoothMoveToPosition(recyclerView, mToPosition);
                }
            }
        });

        mAdapter.setOnItemClickListener(new HoverItemAdapter2.OnItemClickListener() {
            @Override
            public void onItemClick(View view, int position) {
                if (position != -1) {
                    smoothMoveToPosition(mRecyclerView, 20);
                }
            }
        });
    }

    //滑动到指定位置,可以保证滑动后的item位于最上面
    private void smoothMoveToPosition(RecyclerView mRecyclerView, int position) {
        //第一个可见位置
        int firstItem = mRecyclerView.getChildLayoutPosition(mRecyclerView.getChildAt(0));
        //最后一个可见位置
        int lastItem = mRecyclerView.getChildLayoutPosition(mRecyclerView.getChildAt(mRecyclerView.getChildCount() - 1));
        Log.i(TAG, "smoothMoveToPosition: first = " + firstItem + ", position = " + position + ", last = " + lastItem);
        if (position < firstItem) {
            //第一种可能:跳转位置在第一个可见位置之前
            mRecyclerView.smoothScrollToPosition(position);
        } else if (position <= lastItem) {
            //第二种可能:跳转位置在第一个可见位置之后
            int movePosition = position - firstItem;
            int top = mRecyclerView.getChildAt(movePosition).getTop();
            mRecyclerView.smoothScrollBy(0, top);
        } else {
            //第三种可能:跳转位置在最后可见项之后
            mRecyclerView.smoothScrollToPosition(position);
            mToPosition = position;
            mShouldScroll = true;
        }
    }

    private void addData() {
        mDatas = new ArrayList<>();
        mDatas.add("0");
        mDatas.add("1");
        mDatas.add("2");
        mDatas.add("3");
        for (int i = 1; i <= 20; i++) {
            mDatas.add("这是第" + i + "条数据");
        }
    }
}

File: customview/src/main/java/com/example/customview/activity/HoverItemActivity2.kt
package com.example.customview.activity

import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBindings
import com.example.customview.R
import com.example.customview.adapter.HoverItemAdapter2
import com.example.customview.databinding.ActivityHoverItem2Binding

/**
 * 封装 Hover Item Activity 2 相关逻辑的类。
 */
open class HoverItemActivity2 : AppCompatActivity() {
    private lateinit var binding: ActivityHoverItem2Binding
    private lateinit var mAdapter: HoverItemAdapter2
    private lateinit var mRecyclerView: RecyclerView
    private lateinit var mDatas: MutableList<String>
    private lateinit var mIncludeView: View
    private var mTypeName: String? = null
    private var mShouldScroll = false
    private var mToPosition = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityHoverItem2Binding.inflate(layoutInflater)
        setContentView(binding.root)
        mIncludeView = binding.include.root
        mRecyclerView = binding.recyclerView
        addData()
        mAdapter = HoverItemAdapter2(mDatas)
        mRecyclerView.adapter = mAdapter
        mRecyclerView.layoutManager = LinearLayoutManager(this)
        mRecyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                super.onScrolled(recyclerView, dx, dy)
                val stickyInfoView = recyclerView.findChildViewUnder((recyclerView.measuredWidth shr 1).toFloat(), 1f)
                if (stickyInfoView != null) {
                    ViewBindings.findChildViewById<TextView>(stickyInfoView, R.id.typeName)?.let {
                        mTypeName = it.text.toString()
                    }
                    Log.i(TAG, "onScrolled: mTypeName = $mTypeName")
                    when (mTypeName) {
                        "第一项" -> mIncludeView.visibility = View.GONE
                        "第二项", "第三项" -> mIncludeView.visibility = View.VISIBLE
                    }
                }
            }

            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                if (mShouldScroll && RecyclerView.SCROLL_STATE_IDLE == newState) {
                    mShouldScroll = false
                    smoothMoveToPosition(recyclerView, mToPosition)
                }
            }
        })

        mAdapter.setOnItemClickListener(object : HoverItemAdapter2.OnItemClickListener {
            override fun onItemClick(view: View, position: Int) {
                if (position != -1) {
                    smoothMoveToPosition(mRecyclerView, 20)
                }
            }
        })
    }

    private fun smoothMoveToPosition(recyclerView: RecyclerView, position: Int) {
        val firstItem = recyclerView.getChildLayoutPosition(recyclerView.getChildAt(0))
        val lastItem = recyclerView.getChildLayoutPosition(recyclerView.getChildAt(recyclerView.childCount - 1))
        Log.i(TAG, "smoothMoveToPosition: first = $firstItem, position = $position, last = $lastItem")
        when {
            position < firstItem -> recyclerView.smoothScrollToPosition(position)
            position <= lastItem -> {
                val movePosition = position - firstItem
                val top = recyclerView.getChildAt(movePosition).top
                recyclerView.smoothScrollBy(0, top)
            }
            else -> {
                recyclerView.smoothScrollToPosition(position)
                mToPosition = position
                mShouldScroll = true
            }
        }
    }

    private fun addData() {
        mDatas = mutableListOf("0", "1", "2", "3").apply {
            for (i in 1..20) {
                add("这是第${i}条数据")
            }
        }
    }

    companion object {
        private const val TAG = "MainActivity"
    }
}
